### PR TITLE
fix(stream): keep unfinished reasoning private

### PR DESCRIPTION
**Pull request by jonathan308**
## Summary

This is the unfinished-reasoning output-handling split requested in #3327.

- Never reclassify an unterminated private reasoning block as visible answer content.
- Report unfinished OpenAI Chat and Anthropic streams as incomplete/length-capped.
- Emit an incomplete Responses terminal event instead of a completed response.
- Reject structured or raw tool-call candidates that occur inside unfinished private reasoning across Chat, Anthropic, and Responses.
- Preserve normal tag-free text, completed reasoning, valid tools, and existing sampling behavior.

## Scope

This PR intentionally contains no EOS masking and no MTP processor-contract changes. It does not alter logits or prevent a model-selected stop; it handles the resulting incomplete protocol safely.

It is stacked on #3327 because the Qwen streamed-tool suppression uses the capability-scoped streaming seam retained there. After #3327 merges, I will restack this branch onto `main` so the diff collapses to the single four-file output-handling commit.

## Validation

Python 3.13 focused suite: **485 passed, 12 deselected** across unfinished-thinking streams, thinking parsing, Qwen tool calling, and end-to-end streaming.

### CODE REVIEW

**Comment by eazerox2** [comment]
This appears to be the same failure class as this PR. I am adding a production-shaped reproduction from Hermes using Qwen3.8-Flash-Next.

## Environment

- Hardware: Apple Silicon Mac Studio M3 Ultra, 96 GB unified memory
- OS: macOS 26.6.2
- oMLX: 0.7.0.dev1
- Model: `Jundot/Qwen3.8-Flash-Next-oQ4e-mtp`
- API: OpenAI-compatible `POST /v1/chat/completions`
- Client: Hermes Agent
- Streaming: enabled
- Thinking: enabled
- Request reasoning setting: `reasoning_effort=medium`

## Observed production behavior

Two Hermes responses produced the following persisted message state:
text
message IDs: 206474, 206493
finish_reason: stop
structured tool_calls: absent
reasoning_content: non-empty
content: non-empty
content.strip() == reasoning_content.strip()
The visible content was byte-equivalent to the reasoning content. Hermes therefore persisted the model's private reasoning as ordinary assistant content and streamed it to Discord. The turn then ended without a useful final answer.

The affected requests were long, tool-heavy conversations. The approximate request/output sizes were:
text
request 1: ~61.8K prompt tokens, ~602 completion tokens
request 2: ~75.8K prompt tokens, ~580 completion tokens
configured output cap: 16,384 tokens
Both responses ended with `finish_reason="stop"` and were far below the output cap. This does not appear to be ordinary `max_tokens` truncation.

## Minimal parser evidence

The installed parser reproduces the classification problem directly:
python
parser = ThinkingParser(start_in_thinking=True)
parser.feed("private reasoning with no closing boundary")
parser.finish()
Observed result:
text
('', 'private reasoning with no closing boundary')
The accumulated unfinished thinking is returned in the visible-content position. The active Qwen chat template can open generation inside
 (1/2)
``.

## Expected behavior

When a prompt-opened private-thinking block reaches EOS without ``:

- unfinished text must not be promoted to visible `content`;
- it should remain private reasoning or be discarded from the user-visible response;
- the response should be marked incomplete/malformed rather than treated as a successful final answer;
- tool-call candidates inside unfinished private reasoning should not be executed or promoted;
- streaming and non-streaming paths should apply the same policy.

Returning an incomplete response is preferable to exposing private reasoning. I am not requesting automatic EOS masking here; this is specifically about fail-closed output classification after the model has already stopped.

## Related reports

This also appears related to:

- #2457 — unfinished thinking re-emitted as visible content on `finish_reason="length"`;
- #2860 — reasoning/content separation leaks on truncation;
- #2649 — prompt-opened thinking with redundant close markers exposes private text;
- #2744/#2745 — prompt-opened thinking and EOS recovery produce no visible content until final recovery;
- #3129 — Qwen3.8 reasoning/repetition behavior involving reasoning being flushed into visible content.

The distinguishing detail in this report is that the downstream response was externally reported as `finish_reason="stop"` while `content` duplicated `reasoning_content`.

No private prompt text, reasoning text, credentials, or session transcript is included here. I can provide a sanitized raw SSE trace or additional parser-level reproduction if useful.
